src/modm/driver/inertial/qmc5883l.hh

```diff
@@ -0,0 +1,231 @@
+
+public:
+  enum class
+  OversampleRate : uint8_t
+  {
+    _512 = 0,
+    _256 = int(Control1::OSR0),
+    _128 = int(Control1::OSR1),
+    _64  = int(Control1::OSR0) | int(Control1::OSR1),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, OversampleRate);
+
+  enum class
+  FullScale : uint8_t
+  {
+    _2G = 0,
+    _8G = int(Control1::RNG0),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, FullScale);
+
+  enum class
+  OutputDataRate : uint8_t
+  {
+    _10Hz  = 0,
+    _50Hz  = int(Control1::ODR0),
+    _100Hz = int(Control1::ODR1),
+    _200Hz = int(Control1::ODR0) | int(Control1::ODR1),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, OutputDataRate);
+
+  enum class
+  Mode : uint8_t
+  {
+    StandBy = 0,
+    Continious = int(Control1::MODE0),
+  };
+
+  MODM_FLAGS_CONFIG(Control1, Mode);
+
+public:
+	struct modm_packed
+  Data
+	{
+		template <class I2cMaster>
+		friend class Qmc5883l;
+  protected:
+    uint8_t data[7];
+
+    template<Register R, uint8_t Offset = uint8_t(R)>
+    int16_t inline getWord() {
+      static_assert(Offset < sizeof data);
+      const auto value = reinterpret_cast<int16_t*>( data + Offset );
+      return modm::fromLittleEndian( *value );
+    }
+  };
+};
+
+
+template <class I2cMaster>
+class Qmc5883l : public Qmc5883lRegisters, public modm::I2cDevice<I2cMaster>
+{
+public:
+  // This flag is supposed to be set from within an ISR. By the time when the
+  // first call to readData() occurs in the program, the DRDY bit will have
+  // been already set, so need to initialize this flag to "true" to enable
+  // first data read.
+  volatile bool dataReady = true;
+
+private:
+  /// @cond
+  Data &data;
+  /// @endcond
+	uint8_t buffer[sizeof data.data];
+
+public:
+	Qmc5883l(Data &data, uint8_t address=0x0d) :
+    modm::I2cDevice<I2cMaster> (address), data(data) {}
+
+  auto inline
+  x() { return data.getWord<Register::DataX_Lsb>(); }
+
+  auto inline
+  y() { return data.getWord<Register::DataY_Lsb>(); }
+
+  auto inline
+  z() { return data.getWord<Register::DataZ_Lsb>(); }
+
+  auto inline
+  status()
+  {
+    return Status_t( data.data[uint8_t(Register::Status)] );
+  }
+
+public:
+  modm::ResumableResult<bool>
+  configure(Mode_t mode, Control1_t control)
+  {
+    control |= mode;
+    return updateRegister(Register::Control1, control.value);
+  }
+
+
+  modm::ResumableResult<bool>
+  configure(Control2_t control)
+  {
+    return updateRegister(Register::Control2, control.value);
+  }
+
+
+  modm::ResumableResult<bool>
+  configureSetReset()
+  {
+    return updateRegister(Register::SetReset, 0x01);
+  }
+
+
+	modm::ResumableResult<bool>
+	updateRegister(Register reg, uint8_t value)
+	{
+		RF_BEGIN();
+
+		buffer[0] = uint8_t(reg);
+    buffer[1] = value;
+		this->transaction.configureWrite(buffer, 2);
+
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+  modm::ResumableResult<bool>
+  readDataBlocking()
+  {
+    RF_BEGIN();
+    RF_WAIT_UNTIL(dataReady);
+    RF_END_RETURN_CALL(readData());
+  }
```

Comment:
Nah, let the user implement that externally, since there may be other ways of doing it.